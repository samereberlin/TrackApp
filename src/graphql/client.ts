import {ApolloClient, InMemoryCache} from '@apollo/client';

const client = new ApolloClient({
  uri: 'https://me-samer-backend.herokuapp.com/graphql',
  cache: new InMemoryCache(),
});

export default client;
