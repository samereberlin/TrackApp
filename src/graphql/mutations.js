import {gql} from '@apollo/client';

export const CREATE_MEASUREMENT = gql`
  mutation CreateMeasurement($weight: Float!, $measuredAt: Date!, $fileId: ID) {
    measurementCreate(
      input: {weight: $weight, measuredAt: $measuredAt, fileId: $fileId}
    ) {
      id
      weight
      photoPublicPath
      measuredAt
      createdAt
    }
  }
`;

export const UPLOAD_PHOTO = gql`
  mutation UploadPhoto($mimeType: String!) {
    fileUploadInitialize(input: {mimeType: $mimeType}) {
      fileId
      uploadUrl
    }
  }
`;
