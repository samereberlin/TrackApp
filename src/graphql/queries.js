import {gql} from '@apollo/client';

export const GET_MEASUREMENTS = gql`
  query GetMeasurements {
    measurements {
      id
      weight
      photoPublicPath
      measuredAt
      createdAt
    }
  }
`;
