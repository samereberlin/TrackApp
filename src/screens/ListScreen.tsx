import React from 'react';
import {
  ActivityIndicator,
  StyleSheet,
  Text,
  TextStyle,
  ViewStyle,
} from 'react-native';
import {StackNavigationProp} from '@react-navigation/stack';
import {useQuery} from '@apollo/client';

import Container from '../components/Container';
import {GET_MEASUREMENTS} from '../graphql/queries';
import {colors, defaultTextStyle} from '../utils/theme';
import Header from '../components/Header';
import MeasurementList from '../components/MeasurementList';
import {MeasurementType} from '../types';
import PlusButton from '../components/PlusButton';

const styles = StyleSheet.create({
  loading: {flex: 1} as ViewStyle,
  error: {...defaultTextStyle, color: colors.error} as TextStyle,
});

interface ListScreenProps {
  navigation: StackNavigationProp<any>;
}

const ListScreen: React.FC<ListScreenProps> = ({navigation}) => {
  const {loading, error, data} = useQuery(GET_MEASUREMENTS);

  let content, usedDates: number[];
  if (loading) {
    content = (
      <ActivityIndicator
        color={colors.buttonBackground}
        size="large"
        style={styles.loading}
        testID="ActivityIndicator"
      />
    );
  } else if (error) {
    content = <Text style={styles.error}>{error}</Text>;
  } else if (data) {
    const sortedItems: MeasurementType[] = [...data.measurements].sort(
      (a, b) => b.measuredAt - a.measuredAt,
    );
    content = <MeasurementList list={sortedItems} />;
    usedDates = sortedItems.map((item: MeasurementType) => item.measuredAt);
  }

  const buttonCallback = () => {
    navigation.navigate('Create Measurement', {usedDates});
  };
  const buttonDisabled = loading || error !== undefined;

  return (
    <Container>
      <Header label="Measurements" />
      {content}
      <PlusButton disabled={buttonDisabled} onPress={buttonCallback} />
    </Container>
  );
};

export default ListScreen;
