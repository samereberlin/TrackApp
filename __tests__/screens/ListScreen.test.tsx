import 'react-native';
import React from 'react';
import {act, create} from 'react-test-renderer';
import {MockedProvider} from '@apollo/client/testing';
import {render, waitForElementToBeRemoved} from '@testing-library/react-native';

import {GET_MEASUREMENTS} from '../../src/graphql/queries';
import ListScreen from '../../src/screens/ListScreen';
import {StackNavigationProp} from '@react-navigation/stack';

const mocks = [
  {
    request: {
      query: GET_MEASUREMENTS,
    },
    result: {
      data: {
        measurements: [
          {
            createdAt: 1,
            id: 'aa',
            measuredAt: 1,
            photoPublicPath: 'https://via.placeholder.com/150',
            weight: 76,
          },
          {
            createdAt: 172800000,
            id: 'cc',
            measuredAt: 172800000,
            photoPublicPath: 'https://via.placeholder.com/150',
            weight: 74,
          },
          {
            createdAt: 86400000,
            id: 'bb',
            measuredAt: 86400000,
            photoPublicPath: 'https://via.placeholder.com/150',
            weight: 75,
          },
        ],
      },
    },
  },
];

describe('ListScreen', () => {
  it('matches the snapshot', async () => {
    const screen = create(
      <MockedProvider mocks={mocks} addTypename={false}>
        <ListScreen navigation={{} as StackNavigationProp<any>} />
      </MockedProvider>,
    );
    expect(screen.toJSON()).toMatchSnapshot();

    // wait for Apollo/GraphQL mock response:
    await act(async () => {
      await new Promise((resolve) => setTimeout(resolve, 0));
    });

    expect(screen.toJSON()).toMatchSnapshot();
  });

  it('renders correctly', async () => {
    const screen = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <ListScreen navigation={{} as StackNavigationProp<any>} />
      </MockedProvider>,
    );
    expect(screen.getByTestId('ActivityIndicator')).toBeTruthy();

    // wait for Apollo/GraphQL mock response:
    await waitForElementToBeRemoved(() =>
      screen.getByTestId('ActivityIndicator'),
    );

    expect(screen.queryByTestId('ActivityIndicator')).not.toBeTruthy();
    expect(screen.getAllByText('Measurements')).toHaveLength(1);
    expect(screen.getAllByText('ï¼‹')).toHaveLength(1);

    const MeasurementItems = screen.getAllByText(/kg/);
    expect(MeasurementItems).toHaveLength(
      mocks[0].result.data.measurements.length,
    );

    // Measurement id: cc, createdAt: 172800000 = 'Saturday 3 January'.
    expect(MeasurementItems[0]).toHaveTextContent('74kg');
    expect(MeasurementItems[0].parent).toHaveTextContent('Saturday 3 January');

    // Measurement id: bb, createdAt: 86400000 = 'Friday 2 January'.
    expect(MeasurementItems[1]).toHaveTextContent('75kg');
    expect(MeasurementItems[1].parent).toHaveTextContent('Friday 2 January');

    // Measurement id: aa, createdAt: 1 = 'Thursday 1 January'.
    expect(MeasurementItems[2]).toHaveTextContent('76kg');
    expect(MeasurementItems[2].parent).toHaveTextContent('Thursday 1 January');
  });
});
