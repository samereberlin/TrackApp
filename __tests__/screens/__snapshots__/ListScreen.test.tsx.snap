// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListScreen matches the snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#f5f7f7",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "stretch",
        "backgroundColor": "#f8fafa",
        "borderBottomColor": "#f1f1f1",
        "borderBottomWidth": 2,
        "flexDirection": "row",
      }
    }
  >
    <Text
      numberOfLines={1}
      style={
        Object {
          "color": "#333333",
          "flex": 1,
          "fontSize": 20,
          "padding": 16,
        }
      }
    >
      Measurements
    </Text>
  </View>
  <ActivityIndicator
    animating={true}
    color="#44b987"
    hidesWhenStopped={true}
    size="large"
    style={
      Object {
        "flex": 1,
      }
    }
    testID="ActivityIndicator"
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "bottom": 32,
        "left": 0,
        "position": "absolute",
        "right": 0,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#44b987",
          "borderRadius": 60,
          "height": 60,
          "justifyContent": "center",
          "opacity": 1,
          "shadowColor": "black",
          "shadowOffset": Object {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0.4,
          "shadowRadius": 4,
          "width": 60,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "white",
            "fontSize": 40,
            "fontWeight": "bold",
          }
        }
      >
        ï¼‹
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;

exports[`ListScreen matches the snapshot 2`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#f5f7f7",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "stretch",
        "backgroundColor": "#f8fafa",
        "borderBottomColor": "#f1f1f1",
        "borderBottomWidth": 2,
        "flexDirection": "row",
      }
    }
  >
    <Text
      numberOfLines={1}
      style={
        Object {
          "color": "#333333",
          "flex": 1,
          "fontSize": 20,
          "padding": 16,
        }
      }
    >
      Measurements
    </Text>
  </View>
  <RCTScrollView
    ListEmptyComponent={
      <Text
        style={
          Object {
            "fontSize": 16,
            "padding": 16,
          }
        }
      >
        Empty list.
      </Text>
    }
    ListFooterComponent={
      <View
        style={
          Object {
            "height": 64,
          }
        }
      />
    }
    data={
      Array [
        Object {
          "createdAt": 172800000,
          "id": "cc",
          "measuredAt": 172800000,
          "photoPublicPath": "https://via.placeholder.com/150",
          "weight": 74,
        },
        Object {
          "createdAt": 86400000,
          "id": "bb",
          "measuredAt": 86400000,
          "photoPublicPath": "https://via.placeholder.com/150",
          "weight": 75,
        },
        Object {
          "createdAt": 1,
          "id": "aa",
          "measuredAt": 1,
          "photoPublicPath": "https://via.placeholder.com/150",
          "weight": 76,
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "paddingLeft": 8,
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "white",
              "borderRadius": 4,
              "flexDirection": "row",
              "margin": 8,
              "padding": 12,
              "shadowColor": "black",
              "shadowOffset": Object {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0.1,
              "shadowRadius": 1,
            }
          }
        >
          <View
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <Text
              numberOfLines={1}
              style={
                Object {
                  "color": "#333333",
                  "fontSize": 16,
                  "padding": 4,
                }
              }
            >
              Saturday 3 January
            </Text>
            <Text
              numberOfLines={1}
              style={
                Object {
                  "color": "#adc1c5",
                  "fontSize": 16,
                  "padding": 4,
                }
              }
            >
              74
              kg
            </Text>
          </View>
          <Image
            source={
              Object {
                "uri": "https://via.placeholder.com/150",
              }
            }
            style={
              Object {
                "height": 55,
                "width": 55,
              }
            }
          />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "white",
              "borderRadius": 4,
              "flexDirection": "row",
              "margin": 8,
              "padding": 12,
              "shadowColor": "black",
              "shadowOffset": Object {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0.1,
              "shadowRadius": 1,
            }
          }
        >
          <View
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <Text
              numberOfLines={1}
              style={
                Object {
                  "color": "#333333",
                  "fontSize": 16,
                  "padding": 4,
                }
              }
            >
              Friday 2 January
            </Text>
            <Text
              numberOfLines={1}
              style={
                Object {
                  "color": "#adc1c5",
                  "fontSize": 16,
                  "padding": 4,
                }
              }
            >
              75
              kg
            </Text>
          </View>
          <Image
            source={
              Object {
                "uri": "https://via.placeholder.com/150",
              }
            }
            style={
              Object {
                "height": 55,
                "width": 55,
              }
            }
          />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "white",
              "borderRadius": 4,
              "flexDirection": "row",
              "margin": 8,
              "padding": 12,
              "shadowColor": "black",
              "shadowOffset": Object {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0.1,
              "shadowRadius": 1,
            }
          }
        >
          <View
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <Text
              numberOfLines={1}
              style={
                Object {
                  "color": "#333333",
                  "fontSize": 16,
                  "padding": 4,
                }
              }
            >
              Thursday 1 January
            </Text>
            <Text
              numberOfLines={1}
              style={
                Object {
                  "color": "#adc1c5",
                  "fontSize": 16,
                  "padding": 4,
                }
              }
            >
              76
              kg
            </Text>
          </View>
          <Image
            source={
              Object {
                "uri": "https://via.placeholder.com/150",
              }
            }
            style={
              Object {
                "height": 55,
                "width": 55,
              }
            }
          />
        </View>
      </View>
      <View
        onLayout={[Function]}
      >
        <View
          style={
            Object {
              "height": 64,
            }
          }
        />
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "alignItems": "center",
        "bottom": 32,
        "left": 0,
        "position": "absolute",
        "right": 0,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#44b987",
          "borderRadius": 60,
          "height": 60,
          "justifyContent": "center",
          "opacity": 1,
          "shadowColor": "black",
          "shadowOffset": Object {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0.4,
          "shadowRadius": 4,
          "width": 60,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "white",
            "fontSize": 40,
            "fontWeight": "bold",
          }
        }
      >
        ï¼‹
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
